File: src/digidoc.cpp
#include "digidoc-server/include/digidoc.hpp"
#include <Container.h>
#include <string>

void initialize()
{
    digidoc::initialize();
}

void terminate()
{
    digidoc::terminate();
}

class DigiDoc::impl
{
    friend DigiDoc;

    std::unique_ptr<digidoc::Container> doc;

    impl(rust::Str path)
        : doc(digidoc::Container::openPtr((std::string)path))
    {
    }
};

DigiDoc::DigiDoc(rust::Str path)
    : impl(new class DigiDoc::impl(path))
{
}

std::unique_ptr<DigiDoc> container_open_ptr(rust::Str path)
{
    return std::unique_ptr<DigiDoc>(new DigiDoc(path));
}

File: src/digidoc.rs
#[cxx::bridge]
mod ffi {
    unsafe extern "C++" {
        include!("digidoc-server/include/digidoc.hpp");

        type DigiDoc;

        pub fn initialize() -> ();
        pub fn terminate() -> ();

        pub fn container_open_ptr(path: &str) -> Result<UniquePtr<DigiDoc>>;
    }
}

pub struct LibDigiDoc;

impl LibDigiDoc {
    pub fn load() -> Self {
        ffi::initialize();
        println!("DigiDoc initialized ...");
        Self
    }
}

impl Drop for LibDigiDoc {
    fn drop(&mut self) {
        ffi::terminate();
        println!("DigiDoc terminated!");
    }
}
